types/datum.go

```diff
@@ -1423,13 +1423,38 @@ func (d *Datum) convertToMysqlFloatYear(sc *stmtctx.StatementContext, target *Fi
 	return ret, err
 }
 
+func (d *Datum) convertStringToMysqlBit(sc *stmtctx.StatementContext) (uint64, error) {
+	bitStr, err := ParseBitStr(BinaryLiteral(d.b).ToString())
+	if err != nil {
+		// This is to be compatible with the previous processing method.
+		return BinaryLiteral(d.b).ToInt(sc)
+	}
+	return bitStr.ToInt(sc)
+}
+
 func (d *Datum) convertToMysqlBit(sc *stmtctx.StatementContext, target *FieldType) (Datum, error) {
 	var ret Datum
 	var uintValue uint64
 	var err error
 	switch d.k {
-	case KindString, KindBytes:
+	case KindBytes:
 		uintValue, err = BinaryLiteral(d.b).ToInt(sc)
+	case KindString:
+		// To solve issue #18681, we need consider true, false, 0, 1.
+		// More situations, we also consider b'0', b'1' and so on.
+		s := BinaryLiteral(d.b).ToString()
+		if target.Flen == 1 {
+			switch strings.ToLower(s) {
+			case "true", "1":
+				uintValue = 1
```

Comment:
In mysql 8.0:

```sql
mysql > select true, b'1', cast(true as binary);
+------+------------+--------------------------------------------+
| true | b'1'       | cast(true as binary)                       |
+------+------------+--------------------------------------------+
|    1 | 0x01       | 0x31                                       |
+------+------------+--------------------------------------------+
1 row in set (0.00 sec)
```

Maybe we should handle it in another place.

Reply:
This sentence of SQL has nothing to do with convert. It can be seen from the following that these values can be converted correctly, but they are returned as binary, so the output to MySQL Client may not be displayed.
```sql
MySQL [test]> select true, HEX(b'1'), HEX(cast(true as binary));
+------+-----------+---------------------------+
| TRUE | HEX(b'1') | HEX(cast(true as binary)) |
+------+-----------+---------------------------+
|    1 | 01        | 31                        |
+------+-----------+---------------------------+
1 row in set (0.001 sec)

MySQL [test]> select true, b'1', cast(true as binary);
+------+------+----------------------+
| TRUE | b'1' | cast(true as binary) |
+------+------+----------------------+
|    1 |     | 1                    |
+------+------+----------------------+
1 row in set (0.001 sec)
```
I think these are two different problems, and it's best to solve it in another PR